### PR TITLE
Feature/ui hydrator++ bugfixes

### PR DESCRIPTION
**Pull request by ajainarayanan**
- Fixes spacing in error messages in the console
- Adds the ability to show a confirmation whenever the user has made some changes in the canvas and switching to templates, importing a config or changing the artifact.
- Fixes left side panel to have one panel open all the time (removes intermediate state of having all open at the same time)

Bamboo build: http://builds.cask.co/browse/CDAP-DRC3509-3

### CODE REVIEW

File: cdap-ui/app/features/hydratorplusplus/controllers/create/leftpanel-ctrl.js [hunk 1]
```diff
@@ -89,17 +80,30 @@ class HydratorPlusPlusLeftPanelCtrl {
...
+            artifactType: this.selectedArtifact.name,
+            data: null
+          });
+        }
+      });
   }
 
   openFileBrowser() {
-    this.$timeout(function() {
+    let foo = () => {
```
Comment (elfenheart) [inline cdap-ui/app/features/hydratorplusplus/controllers/create/leftpanel-ctrl.js]: should we rename this to a more meaningful function name?

Comment (ajainarayanan) [inline cdap-ui/app/features/hydratorplusplus/controllers/create/leftpanel-ctrl.js]: oops. Sorry missed renaming it.

File: cdap-ui/app/directives/group-side-panel/group-side-panel-ctrl.js [hunk 2]
```diff
@@ -17,22 +17,7 @@
...
-      }
-
-      angular.forEach(this.groups, function (g) {
-        g.expanded = false;
-      });
-
-      group.expanded = true;
-      this.openedGroup = group.name;
-    };
+    this.openedGroup = this.groups[0].name;
```
Comment (russorat) [inline cdap-ui/app/directives/group-side-panel/group-side-panel-ctrl.js]: is this.groups guaranteed to exist and not be empty?

File: cdap-ui/app/features/hydratorplusplus/controllers/create/create-studio-ctrl.js [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 class HydratorPlusPlusStudioCtrl {
   // Holy cow. Much DI. Such angular.
-  constructor(HydratorPlusPlusLeftPanelStore, HydratorPlusPlusConfigActions, $stateParams, rConfig, $rootScope, $scope, DetailNonRunsStore, HydratorPlusPlusNodeConfigStore, DAGPlusPlusNodesActionsFactory, HydratorPlusPlusHydratorService, HydratorPlusPlusConsoleActions) {
+  constructor(HydratorPlusPlusLeftPanelStore, HydratorPlusPlusConfigActions, $stateParams, rConfig, $rootScope, $scope, DetailNonRunsStore, HydratorPlusPlusNodeConfigStore, DAGPlusPlusNodesActionsFactory, HydratorPlusPlusHydratorService, HydratorPlusPlusConsoleActions, rSelectedArtifact, rArtifacts) {
```
Comment (russorat) [inline cdap-ui/app/features/hydratorplusplus/controllers/create/create-studio-ctrl.js]: this line is crazy long

File: cdap-ui/app/features/hydratorplusplus/controllers/create/create-studio-ctrl.js [hunk 4]
```diff
@@ -29,9 +29,17 @@ class HydratorPlusPlusStudioCtrl {
       HydratorPlusPlusNodeConfigStore.reset();
       HydratorPlusPlusConsoleActions.resetMessages();
     });
-
+    let getValidArtifact = () => {
+      let isValidArtifact;
+      if (rArtifacts.length) {
+        isValidArtifact = rArtifacts.filter(r => r.name === rSelectedArtifact);
+      }
+      return isValidArtifact.length? isValidArtifact[0]: rArtifacts[0];
```
Comment (russorat) [inline cdap-ui/app/features/hydratorplusplus/controllers/create/create-studio-ctrl.js]: not sure about the javascript coding style here but should there be spaces before the question marks and colons?

File: cdap-ui/app/features/hydratorplusplus/templates/create/popovers/canvas-overwrite-confirmation.html [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<!--
+  Copyright © 2015 Cask Data, Inc.
```
Comment (russorat) [inline cdap-ui/app/features/hydratorplusplus/templates/create/popovers/canvas-overwrite-confirmation.html]: 2016

File: cdap-ui/app/features/hydratorplusplus/controllers/create/create-studio-ctrl.js (hunk 3 above)
Comment (ajainarayanan) [inline cdap-ui/app/features/hydratorplusplus/controllers/create/create-studio-ctrl.js]: Yeah.. Thats angular's dependency injection. This particular controller requires that many dependencies. This is not long. There are other controllers that require much more dependencies. The good part is we (user code) doesn't call this function at all. Angular makes sure it calls this function with all arguments.

**Comment by elfenheart** [comment]
:+1: LGTM

**Pull request merged**